OSSMDOC-125 and 131, configure external Elasticsearch.

JStickler opened the pull request:
This PR addresses both [OSSMDOC-125](https://issues.redhat.com/browse/OSSMDOC-125) (Document external Elasticsearch) and [OSSMDOC-131](https://issues.redhat.com/browse/OSSMDOC-131) (split out auto-provision and self-provisioned Elasticsearch options).

JStickler commented:
@objectiser , @jpkrohling,  Could you guys look this over and make sure I’ve split things out correctly?  I wasn’t sure if the esIndexCleaner options should go both places or not.

Also, have I got the restrictions correct for each?  Are they any additional restrictions for external Elasticsearch?

objectiser commented on modules/jaeger-config-storage.adoc:
```diff
@@ -157,6 +118,152 @@ Minimum deployment = 16Gi*
+|The fully-qualified domain name of the Elasticsearch server.
+|`http://elasticsearch.<namespace>.svc:9200`
+|es:
+  max-num-spans:
```
Note, in the upstream docs "(deprecated, will be removed in release v1.21.0. Please use es.max-doc-count). The maximum number of spans to fetch at a time per query in Elasticsearch. The lesser of es.max-num-spans and es.max-doc-count will be used if both are set."

JStickler replied:
So should we leave max-num-spans in with the deprecation notice,?  Or take it out and only document es.max-doc-count?

objectiser replied:
@jpkrohling thoughts?
I think probably better to include the deprecation notice, as although the operator will handle the upgrade case, users may still try to use an old CR and deploy with 1.20GA.

objectiser commented on modules/jaeger-config-storage.adoc:
```diff
@@ -157,6 +118,152 @@ Minimum deployment = 16Gi*
 |
 
+4+|Elasticsearch index and index cleaner configuration options
+
```
Index cleaner is relevant to auto and self provisioned elasticsearch, so should be moved into a common section.

JStickler replied:
But wouldn't you specify this either under  "storage/elasticsearch" or "storage/options/es"?  That was why I had it in both tables.  I'm worried that if I put in under a common section that it might be confusing where to put in the YAML.

objectiser replied:
No, the index cleaner configuration is directly under the `storage` node: https://www.jaegertracing.io/docs/1.20/operator/#elasticsearch-index-cleaner-job

objectiser commented on modules/jaeger-config-storage.adoc:
This should be removed - the `server-urls` option is only specified for self provisioned ES clusters and does not belong under the `storage.elasticsearch` node, but under `storage.options.es`,

JStickler commented on modules/jaeger-config-storage.adoc:
```diff
@@ -157,6 +118,152 @@ Minimum deployment = 16Gi*
+|Cron expression
+|"55 23 * * *"
+
+|esRollover:
```
Yes, it's easy to pull things out.

JStickler commented on modules/jaeger-config-storage.adoc:
```diff
@@ -157,6 +118,152 @@ Minimum deployment = 16Gi*
+[id="jaeger-config-external-es_{context}"]
+== Connecting to an existing Elasticsearch instance
+
+Jaeger also allows you to use an existing (self-provisioned) Elasticsearch cluster for storage.  You do this by specifying the URL of the existing cluster as the`spec:storage:es:server-urls` value in your configuration.
```
Will update.

jpkrohling commented on modules/jaeger-config-storage.adoc:
```diff
@@ -157,6 +118,152 @@ Minimum deployment = 16Gi*
+
+.Restrictions
+
+* There can be only one Jaeger with self-provisioned Elasticsearch instance per namespace.
```
As far as I understand, this is a limitation of the ES operator, but we need this documented somewhere.

objectiser commented:
@JStickler In future would be better to avoid squashing the commits until merge - do you have a summary of the changes that were made?

JStickler commented:
Updated based on review comments.

- Moved exIndexCleaner options up to general storage parameters table
- Added rows for max-doc-count
- Added deprecation notices to max-num-spans

JStickler commented on modules/jaeger-config-storage.adoc:
```diff
@@ -422,9 +410,15 @@ The following configuration parameters are for an _external_ Elasticsearch insta
 |es-archive:
   max-num-spans:
-|The maximum number of spans to fetch at a time, per query, in Elasticsearch.
+|[*Deprecated* - Will be removed in a future release, use `es.max-doc-count` instead.] The maximum number of spans to fetch at a time, per query, in Elasticsearch.
```
Thanks, that was a copy/paste error.

JStickler commented on modules/jaeger-config-storage.adoc:
```diff
@@ -33,6 +33,29 @@ Memory storage is only appropriate for development, testing, demonstrations, and
 |
 |N/A
+
+4+|Elasticsearch index and index cleaner configuration options
```
I'm going to leave it where it is for now, we can always move it later if we add content.

JStickler commented on modules/jaeger-config-storage.adoc:
```diff
@@ -157,6 +134,136 @@ Minimum deployment = 16Gi*
+    type: elasticsearch
+    elasticsearch:
+      nodeCount: 1
+      storage: # <1>
```
Not sure which option "this" refers to, but we do already link to the Elasticsearch Operator docs from the Jaeger docs.

objectiser replied:
Not sure now :smile:

JStickler commented:
This update

- Corrects copy/paste error for es-archive:max-number-spans
- Corrects descriptions of index cleaner and index options in tables
- Added missing es:sniffer-tle-enabled and es-archive:sniffer-tle-enabled

longmuir commented:
This PR looks like it only covers updating the standalone Jaeger docs. Do we have another update in the "Service Mesh", "Customizing the Installation" section to indicate to advise on how to configure an external ElasticSearch instance? I don't think we want to duplicate content, but this is a feature Service Mesh users have asked for, so I think it'd be good to at least call it out there and link to the external ES docs from the main Service Mesh section. Ignore this if it's taken care of already :).

JStickler commented:
@longmuir , this PR is for configuring and external Elasticsearch.  Configuring an external Jaeger is covered on a separate issue (OSSMDOC-132) and will be a separate PR.

neal-timpe commented on modules/jaeger-config-storage.adoc:
I think this should be a separate table. I think if the table needs explanation halfway through, they're different enough to be in separate tables. Also, it's better for accessibility. But I also think two tables with three rows is more consumable than one with six.

longmuir commented:
@JStickler I was referring to an external ES (not external Jaeger). What I meant is this PR looks like it's only updating the Jaeger section of the docs, so someone installing OSSM (From here: https://docs.openshift.com/container-platform/4.6/service_mesh/v1x/customizing-installation-ossm.html), will have no way of knowing that it's possible to use an external ElasticSearch with Service Mesh (which is really just for Jaeger). This is as much an OSSM feature as it is a Jaeger feature.

I want to make sure that someone installing OSSM is able to find their way to this information. I think we should add something to the page I linked above to point to these docs on how to configure an external ElasticSearch with Jaeger in OSSM, to make sure that an OSSM user is aware that this is possible.

neal-timpe commented on modules/jaeger-config-storage.adoc:
```diff
@@ -157,6 +134,142 @@ Minimum deployment = 16Gi*
+
+The following configuration parameters are for an _external_ Elasticsearch instance, that is, an instance that was not created using the Elasticsearch Operator. You specify configuration options for external Elasticsearch under `spec:storage:options:es` in your configuration file.
+
+.ES configuration parameters
```
I think this table should be broken into several tables also.

neal-timpe commented on modules/jaeger-config-storage.adoc:
```diff
@@ -491,18 +573,24 @@ spec:
 
-<1> Persistent storage configuration. In this case AWS `gp2` with `5Gi` size. When no value is specified, Jaeger uses `emptyDir`. The Elasticsearch Operator provisions `PersistentVolumeClaim` and `PersistentVolume` which are not removed with Jaeger instance. You can mount the same volumes if you create a Jaeger instance with the same name and namespace.
+----
+<1> Url to Elasticsearch service running in default namespace.
```
Url/URL

JStickler commented:
@longmuir, as far as I can tell, service mesh doesn't yet fully support external Elasticsearch (all the work under MAISTRA-1087 is not yet complete, and MAISTRA-1065 - Support for external Elasticsearch is still open).

JStickler commented:
Latest update breaks tables into smaller tables per Neal's suggestion (and Gary's earlier comment).

neal-timpe commented:
/cherry-pick enterprise-4.6

neal-timpe commented:
/cherry-pick enterprise-4.7
